sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+namespace Azure.Core
+{
+    internal class OperationInternal<TResult>
+    {
+        private readonly TimeSpan _defaultPollingInterval;
+
+        private readonly IOperation<TResult> _operation;
+
+        private TResult _value;
+
+        public OperationInternal(string operationTypeName,
+            ClientDiagnostics clientDiagnostics,
+            IOperation<TResult> operation,
+            TimeSpan? defaultPollingInterval = null) : this(operationTypeName, clientDiagnostics, defaultPollingInterval)
+        {
+            _operation = operation;
+        }
+
+        protected OperationInternal(string operationTypeName,
+            ClientDiagnostics clientDiagnostics,
+            TimeSpan? defaultPollingInterval)
+        {
+            OperationTypeName = operationTypeName;
+            ClientDiagnostics = clientDiagnostics;
+            _defaultPollingInterval = defaultPollingInterval ?? TimeSpan.FromSeconds(1);
+        }
+
+        public bool HasValue { get; private set; }
+
+        public bool HasCompleted { get; set; }
```

Comment:
There are clients that, under certain conditions, can already set `HasCompleted` to `true` during construction. Making it settable like this could cause some race conditions if not used correctly. For instance, if the client has the following code:
```cs
HasCompleted = true;
Value = ...;
```

A user could end up doing:
```cs
if (operation.HasCompleted)
{
  var value = operation.Value; // may return null because we could have not set Value yet
}
```
So developers need to avoid updating `HasCompleted` before `Value`.

We could:
- Prevent clients from setting `HasCompleted` manually. They would need to work around it in their Operation implementation with private fields (`public bool HasCompleted => _hasCompleted` instead of relying on `OperationInternal`).
- We can hide the setter, but make it possible to update the property through a separate method, something like `SetOperationComplete(TResult value)`. This impacts discoverability, though. I also wonder if there's any extra scenario in which we may want to set `HasCompleted = true` without having a value (like when the operation fails).

Reply:
Wonder if we can have `SetState(OperationState<> )`. That might handle multiple scenarios at once. As for discoverability we'll have to see... I'd rather all value updates were going through the same logic.